### DISCUSSION TITLE
用 git 指令讓你交作業更帥氣（初學者友善）

### DISCUSSION THREAD

**Discussion by tonyYZL** [1101-GS4719-C/Python-Community-GS4719-C#1186] [General]
這禮拜出作業了～
剛剛花了一些時間速速寫完，要上傳的時候覺得一題一題編輯有點慢，我想要再快一點

**Git 是你的救星！**

那要怎麼用勒？？

## Step1. 安裝 git （有安裝過的直接看Step2）
這邊只講 mac 的安裝方式，用 windows 系統的朋友抱歉，~~自己想辦法~~我手邊沒有 windows 的電腦讓我測，可以先到 [Git官網](https://git-scm.com/) 看看

1. 打開終端機，打這條指令下載 Homebrew（mac 的套件管理工具，大推🦵）
```
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

2. 安裝 git，一樣在終端機輸入
```
brew install git
```

3. 最後用這條指令確認是否安裝成功，如果有顯示版本號碼的話就表示成功了！
```
git --version
```
<img width="281" alt="image" src="https://user-images.githubusercontent.com/87929219/228597402-1f0f0be4-17dc-4294-96b4-9d5403cf7d15.png">

## Step2. 把專案 clone 到自己的電腦
clone 就是把整個專案下載下來，方法是...

1. 在 Github 上打開作業，也就是`1112-python-hw1-你的名字`這個專案
2. 看到綠色按鈕上面寫著 Code，給它按下去
3. 複製這個連結👇
<img width="430" alt="image" src="https://user-images.githubusercontent.com/87929219/228599501-6f270c5d-6696-47db-8459-2e72fa72630c.png">

4. 再打開終端機，先切到你想要它在的資料夾，我選桌面
```
cd 資料夾路徑
```
5. 接著輸入以下指令

```
git clone 你複製的那個連結
```
<img width="547" alt="image" src="https://user-images.githubusercontent.com/87929219/228601080-e9d7670e-ca9f-40ae-847b-3967a7933147.png">

成功！這時候你的桌面（或你選的位置）應該會出現跟專案同名的資料夾～
## Step3. 寫作業
~~自己寫~~

## Step4. 交作業
前面廢話一堆，重點來了🚨

假如你用 VS Code 寫，完成後左邊側邊欄應該會長這樣。`M`表示這個檔案的內容有變動
<img width="199" alt="image" src="https://user-images.githubusercontent.com/87929219/228603975-e37f3207-6aed-43e6-a9c1-dea3ad459a95.png">

接著在 VS Code 裡打開終端機，依序輸入三行指令
1. 將所有檔案加入索引（類似暫存區，還沒正式提交到專案裡）
```
git add .
```
2. 提交到本地專案，引號內可以寫一些訊息，讓別人知道你這次提交做了什麼
```
git commit -m "hw hand in"
```
3. 提交到遠端專案（也就是Github畫面上看到的那個），大功告成！
```
git push
```
這時候系統會自動幫你改作業，綠色勾勾就表示沒問題，如果是紅色叉叉的話可能是程式有寫錯喔
<img width="815" alt="image" src="https://user-images.githubusercontent.com/87929219/228610292-e351d329-d7f3-405d-9740-4e86f30ae7e5.png">

**之後如果要再上傳檔案的話，重複以上三行指令就可以了**

熟悉 git 真的很方便，假如老師下次出10題，你還是可以很優雅地把作業們上傳上去

這篇給想嘗試用 git 的人參考，同時也是讓自己複習一下，如果有錯誤的話歡迎各位指正，謝謝大家～

**Comment by xxrjun** [comment]
windows的同學可以安裝[Git for Windows](https://gitforwindows.org/)，這個會一起安裝Git BASH(Git專門的終端機介面)，裝完即可使用指令操作Git，我自己是推薦搭配[Windows Terminal](https://apps.microsoft.com/store/detail/windows-terminal/9N0DX20HK701?hl=zh-tw&gl=tw)比較方便~

不想打指令的話可以使用GUI介面工具，如[GitHub Desktop](https://desktop.github.com/)或是比較進階的[SourceTree](https://www.sourcetreeapp.com/)。或是直接於VS Code中操作也是我很喜歡的做法之一。

修改完檔案打開Source control的儀錶板
- 於Chages中將修改過的檔案加入暫存區(staging area) = 指令 `git add <filename>`
- 於上方欄位輸入commit message並按下enter鍵，會讓Staged Changes中的檔案提交至localrepo = 指令 `git commit -m "<your commit message>"`
- 按下Sync Changes，會同時進行遠端拉取(pull)與推送(push) = 指令 `git pull` + `git push`

![image](https://user-images.githubusercontent.com/40348319/228620730-4a9b1514-4747-487b-8853-82221d9ee232.png)

![image](https://user-images.githubusercontent.com/40348319/228621918-03fdc02e-13d8-4971-af0d-f121cc321907.png)


補充一下: 喜歡從terminal開VS Code的同學可以於VS Code中按下 `Ctrl + Shift + P` 輸入 `shell command` 找到 `Shell Command: Install 'code' command in PATH` 並按下Enter。這樣以後都可以從Terminal進到你要開啟的資料夾中輸入 `code .` 即可使用VS Code開啟該專案。

**Reply by tonyYZL** [reply]
謝謝同學幫忙補充！
（話說你的 README 自介好colorful，超美）

**Comment by Lipunpun** [comment]
肯定要推一下的吧，光是這個標題交作業還可以帥氣，我就給100分，如果我是老師看到這樣交作業的，眼淚一定都快滴出來了。

**Reply by chia-kai-chang** [reply]
火箭已送出~ 🚀🚀🚀

**Comment by JohnnyXiao1024** [comment]
太感動了.....身為幾乎第一次使用git的菜雞，這篇文章真的受益良多
謝謝同學的無私分享，我也送出一個火箭～～（雖然我的火箭沒有用就是了wwww

**Comment by Elaine2899** [comment]
好強...... 身為真的第一次使用github的菜雞，這次交作業遇到滿滿困難，希望之後慢慢更熟悉github與git，也想試試同學分享的這個大好方法!! 希望我會成功www

**Comment by tower20021202** [comment]
感謝大神，這對於不太會用git的人來說真的太重要了，不然作業就要交不出來了

**Comment by YoYooHuang** [comment]
太強啦！我本來是用jupyter lab寫，寫一寫再提交，雖然也是很快，但這個感覺帥多了，我下次會嘗試看看的！

**Comment by RogelioKG** [comment]
真的很想學習用 git 指令🥹 commit & push，
還有關於 branch 和 main 之類的部分。
總覺得要好好學起來才是有真的在用 Github。
希望這次暑假能好好地掌握這些技能。

**Comment by KH1126** [comment]
最近正在通過學習git指令來學習版本控制，通過進行版本控制來協作專案真的很好用。
我認為學會git指令對專題分工製作有非常大的幫助。

**Comment by benbebobo149** [comment]
對我來說，我覺得在推廣git的使用時，更推薦使用GitHub Desktop或是vscode內建的git功能。相比於純粹使用git指令，這些工具更加直觀且易於上手，對於初學者來說更友好。雖然了解git指令也很重要，但重點在於理解git的核心概念和運作原理。無論是使用哪種方式，只要能夠領悟到git的精髓，就能夠更有效地應用它來管理版本控制。感謝同學的分享，讓我們能夠討論和分享有關程式開發工具的使用建議。

**Comment by OG-Matcha** [comment]
我覺得 Git 確實能夠在和他人合作進行大型專案時發揮很好的協作功能，像樓上說的 GitHub Desktop 可以直接用 GUI 的方式去進行 Git 操作，相當的方便。但能夠知道每個按鍵背後的實際指令對於觀念理解也是相當重要的，以上是我個人的見解。

**Comment by sunset513** [comment]
### 其他的方法
就我個人的使用經驗，當你使用 GitHub 時，除了使用指令行工具來進行各種操作外，你也可以透過 vscode 界面來進行相關操作。例如，你可以透過點擊按鈕或連結來進行提交程式碼、建立新的分支、發起合併請求、審查他人的程式碼等。我自己就常常這樣用都快忘記怎麼打指令了

<img width="200" alt="image" src="https://github.com/NCU-GS4719-Python/Python-Community/assets/100106590/3caab503-682e-4f3e-812d-9838fb9fdbe5">

**Comment by YifenLin** [comment]
挖歐，我現在才挖到這個寶藏QQ，之前有短暫實習一個半月，發現跟別人協作寫progect真的會git超超超及重要，認真覺得我很需要好好學好這個技能，雖然網路上資訊很多，不過資訊太雜學習很沒有系統性，組織架構凌亂，不知道有沒有git大神能夠分享學習方法呢?個人覺得還是要有一個機會實際操作比較不會學過就忘哈哈

**Comment by yyujimmy** [comment]
這篇文章提供了很棒的 Git 使用指南！這對於那些剛開始學習 Git 或者想要更深入了解 Git 的人來說都是非常有用的。而你的心情排名也非常貼切，看來你對於這個主題是非常有熱情的呢！

這篇文章的內容很清晰明瞭，對於每個步驟都有詳細的解釋，讓讀者能夠迅速上手。另外，回覆中提到了安裝 Git for Windows 以及使用 GUI 工具的方法，這些補充訊息也相當有幫助。

感謝你分享這篇文章，對於想要學習 Git 的人來說，這將是一個極好的起點！😊

**Comment by William08290829** [comment]
感謝同學分享，有感受到版主的用心~ 我原本自己已經會使用git了，但看完這篇文章還是學到不少，所以也謝謝版主介紹讓我又學到更多~
